### ISSUE TITLE
A way to store backups in a non-cloud storage?

### ISSUE THREAD

**Issue by x-yuri** [pgbackrest/pgbackrest#1754]
From what I can see, pgBackRest needs to be installed on a repository host. But with e.g. Hetzner's [Storage Box][a] you basically have only SFTP access.

I'm now thinking of keeping the repository locally and `rsync`ing it after every backup. But there're also WAL files. And there's no way to know when a new WAL file was added. Which means I have to compromise and just `rsync` occasionally (which increases RPO).

Also, I'm concerned with the fact that the WAL files can change while I `rsync`. Can that leave me with broken WAL files? Well, it can probably affect only the last file. Which supposedly means that in the worst case I'll be unable to PITR to a point in the corresponding time range.

In other words, with such a storage the only downside is bigger RPO.

Does my solution sounds good to you? Any suggestions?

[a]: https://docs.hetzner.com/robot/storage-box/access/access-ssh-rsync-borg/

**Comment by pgstef** [comment]
You have 2 ways : using a dedicated repository host, or using a directly attached storage in the supported [repository types](https://pgbackrest.org/configuration.html#section-repository/option-repo-type).

Using rsync is imo not a good idea (too many data could be lost). In the supported repo types, you don't just have S3 (or all the S3 compatible storage like MinIO), Azure or gcs. You also have nfs or cifs. 
I see in the docs that what you plan to use could provide [cifs](https://docs.hetzner.com/robot/storage-box/access/access-samba-cifs). That's probably way better than using rsync...

**Comment by sfrost** [comment]
The right answer to this is to implement an sftp driver in pgbackrest.  We'd welcome such an effort but we don't have any immediate plans to implement it in the near-term.  Storage drivers are very clear and separate in the code base though and there's a number of them already to use as examples.  I'd suggest you take a look at that and perhaps you can be the one who writes a patch to implement sftp support for pgbackrest.  We'd be happy to review patches.

**Comment by x-yuri** [comment]
> Using rsync is imo not a good idea (too many data could be lost).

Could be lost because I can't run `rsync` after every change? The data since the last time I ran `rsync` will be lost?

> The right answer to this is to implement an sftp driver in pgbackrest.

Is anything wrong with cifs? Should I expect any issues?

> I'd suggest you take a look at that and perhaps you can be the one who writes a patch to implement sftp support for pgbackrest.

As much as I'd like to... I'm not a C programmer, although I have some basic knowledge. I took a look, and... it looks too complex. I don't have that much time. By the way, is there a `remote` driver? Or those are just some helper files?

**Comment by sfrost** [comment]
Using rsync against a pgbackrest repo should generally be fine, it's specifically intended to be able to be backed up by traditional backup tools.  If there is a reason why that doesn't work, we'd certainly like to know.

Note that you absolutely can *not* use rsync to backup a PG data directory directly, that's a whole different thing...  But a pgbackrest repo should be fine.

**Comment by x-yuri** [comment]
About CIFS... As far as I can see, `chown` and `chmod` are noop's. Is it okay with pgBackRest? All the files and dirs are visible as 0755, in contrast to SFTP where they differ. And all of them (supposedly) owned by `root:root` (which is probably okay). Is this specifics of this particular (backup) server?

In a test repo all files has 0640, dirs 0750. Owned by `root:root`. Not sure if that can be taken as an invariant.

I guess that should work for everything but restore. Before restore one should copy the repo, fix permissions, unmount, move the copy back, restore. Which is kind of cumbersome. And makes me think that `rsync` is better after all. Although with `rsync` I might lose the changes since the last time I ran `rsync`.

What do you think?

**UPD** The way I mount it:

```
$ mount.cifs -o cred=creds.txt //domain/backup /mnt
```

On second thought, I can probably make use of the `file_mode`, `dir_mode` options (`file_mode=0640,dir_mode=0750`). What do you say?

**Comment by x-yuri** [comment]
> Although with `rsync` I might lose the changes since the last time I ran `rsync`.

The effects of which can be reduced, in my view, by also running a standby (standby for a disk failure, backups for PITR).

**Comment by silenius** [comment]
Interesting topic, I'm busy to configure pgbackrest with multiple repositories (one local and one remote) and was wondering the same. I have a dedicated Hetzner server (FreeBSD 13.1 and ZFS) and a wireguard link to our local routers. I was planning to setup NFS (V4) but was wondering about the POSIX compliance. The pgbackrest documentation mentions POSIX repository but as NFS is not POSIX compliant I was wondering if I could risk corruption and/or lose data?

**Comment by dwsteele** [comment]
> On second thought, I can probably make use of the file_mode, dir_mode options (file_mode=0640,dir_mode=0750). What do you say?

This is probably the better way to go.

**Comment by dwsteele** [comment]
> The pgbackrest documentation mentions POSIX repository but as NFS is not POSIX compliant I was wondering if I could risk corruption and/or lose data?

We support NFS but prefer the same settings as PostgreSQL (as our docs state): https://www.postgresql.org/docs/current/creating-cluster.html#CREATING-CLUSTER-FILESYSTEM

**Comment by x-yuri** [comment]
> All the files and dirs are visible as 0755, in contrast to SFTP where they differ. And all of them (supposedly) owned by root:root (which is probably okay).

This supposedly doesn't matter, since permissions and ownership in a data dir and a repo differ anyway. pgBackRest either sets the correct permissions on restore or relies on umask. But:

```
2022-05-20 17:38:00.003 P00   INFO: backup command begin 2.36: --exec-id=33-1ec2d90c --log-level-console=info --pg1-host=db --pg1-path=/var/lib/postgresql/data --repo1-path=/mnt/pgbackrest --repo1-retention-full=2 --stanza=db --type=full
2022-05-20 17:38:00.829 P00   INFO: execute non-exclusive pg_start_backup(): backup begins after the next regular checkpoint completes
2022-05-20 17:38:01.231 P00   INFO: backup start archive = 000000010000000000000002, lsn = 0/2000060
2022-05-20 17:38:30.843 P00   INFO: execute non-exclusive pg_stop_backup() and wait for all WAL segments to archive
2022-05-20 17:38:31.044 P00   INFO: backup stop archive = 000000010000000000000002, lsn = 0/2000170
2022-05-20 17:38:31.119 P00   INFO: check archive for segment(s) 000000010000000000000002:000000010000000000000002
2022-05-20 17:38:31.296 P00   INFO: new backup label = 20220520-173800F
ERROR: [041]: unable to create symlink '/mnt/pgbackrest/backup/db/latest' to '20220520-173800F': [95] Not supported
2022-05-20 17:38:31.522 P00   INFO: backup command end: aborted with exception [041]
```

Which apparently means that CIFS is not an option at least in my case. As such `rsync` is probably the only option. Unless we consider a dedicated host, but I'd like to focus on storage here. And with `rsync` data after last `rsync` will apparently be unavailable for PITR.

Also, a strange thing about CIFS is that I couldn't delete the repo in `mc` in one go. At some point it stopped refusing to delete the files because "Directory not empty." So I did "Skip all," then delete again, and after a couple of times the directory was gone.

**Comment by hunleyd** [comment]
> ERROR: [041]: unable to create symlink '/mnt/pgbackrest/backup/db/latest' to '20220520-173800F': [95] Not supported

This indicated repo1-type was not set to cifs

**Comment by x-yuri** [comment]
Good point. With `repo-type=cifs` it works. Until the moment it has to delete files (retention):

```
2022-05-22 02:10:00.003 P00   INFO: backup command begin 2.36: --exec-id=67-5e36429d --log-level-console=info --pg1-host=db --pg1-path=/var/lib/postgresql/data --repo1-path=/mnt/pgbackrest --repo1-retention-full=2 --repo1-type=cifs --stanza=db --type=full
2022-05-22 02:10:00.840 P00   INFO: execute non-exclusive pg_start_backup(): backup begins after the next regular checkpoint completes
2022-05-22 02:10:01.243 P00   INFO: backup start archive = 000000010000000000000006, lsn = 0/6000028
2022-05-22 02:10:30.525 P00   INFO: execute non-exclusive pg_stop_backup() and wait for all WAL segments to archive
2022-05-22 02:10:30.726 P00   INFO: backup stop archive = 000000010000000000000006, lsn = 0/6000138
2022-05-22 02:10:30.760 P00   INFO: check archive for segment(s) 000000010000000000000006:000000010000000000000006
2022-05-22 02:10:31.142 P00   INFO: new backup label = 20220522-021000F
2022-05-22 02:10:31.514 P00   INFO: full backup size = 23.6MB, file total = 979
2022-05-22 02:10:31.514 P00   INFO: backup command end: completed successfully (31511ms)
2022-05-22 02:10:31.514 P00   INFO: expire command begin 2.36: --exec-id=67-5e36429d --log-level-console=info --repo1-path=/mnt/pgbackrest --repo1-retention-full=2 --repo1-type=cifs --stanza=db
2022-05-22 02:10:31.524 P00   INFO: repo1: expire full backup 20220522-020739F
2022-05-22 02:10:31.578 P00   INFO: repo1: remove expired backup 20220522-020739F
ERROR: [061]: repo1: unable to remove path '/mnt/pgbackrest/backup/db/20220522-020739F/pg_data/global': [39] Directory not empty
ERROR: [104]: expire command encountered 1 error(s), check the log file for details
2022-05-22 02:10:31.683 P00   INFO: expire command end: aborted with exception [104]
```

Here's what the mount looks like:

```
//....your-storagebox.de/backup on /mnt type cifs (rw,relatime,vers=3.1.1,cache=strict,username=...,uid=0,noforceuid,gid=0,noforcegid,addr=...,file_mode=0600,dir_mode=0700,soft,nounix,serverino,mapposix,rsize=4194304,wsize=4194304,bsize=1048576,echo_interval=60,actimeo=1)
```

It probably experiences the same difficulties as `mc` above. Is there anything I can check?

**Comment by x-yuri** [comment]
The issue with removing files seems to be [Alpine-specific][a]. Or supposedly you need a lot of files for it to manifest itself with other distros/`glibc`. One might probably say it's a pgBackRest's bug.

Also, I've created a [repo][b] with examples of using pgBackRest under `docker`. Constructive criticism is welcome.

> We support NFS but prefer the same settings as PostgreSQL (as our docs state): https://www.postgresql.org/docs/current/creating-cluster.html#CREATING-CLUSTER-FILESYSTEM

Which `repo-type` then? `posix`?

[a]: https://gitlab.alpinelinux.org/alpine/aports/-/issues/13856
[b]: https://github.com/x-yuri/pgbackrest-docker

**Comment by dwsteele** [comment]
> One might probably say it's a pgBackRest's bug.

Not sure how `readdir()` returning the wrong results equals pgbackrest bug, but this is not the first time we have seen something like this. Interesting that there is a repro but the `readdir()` issue does not seem to have been fixed.

We may need to implement some sort of retry to handle this. So far we have resisted doing so because we want the remove to fail in the case that another process is writing files into the directory.

> Which repo-type then? posix?.

Yes.

**Comment by x-yuri** [comment]
> Not sure how readdir() returning the wrong results equals pgbackrest bug

I think it's a controversial topic. The explanation supposedly is that when you iterate over files and delete them, there's buffering in place. So, you ask for the first "page" (a list of files), delete them, then ask for the second, but owing to the first page having been deleted what you get is what was initially the third page. As a result, the second page (in terms of the initial list) is left undeleted, and deleting the directory fails (since it's not yet empty). And `pgbackrest backup`, in its turn, fails as well.

Now, who's to blame? The kernel part (supposedly) for not hiding this from programs (but then what? introduce snapshots and transactions?)? Or programs for not taking this into account? Hard to say (for me). I don't insist that the fault is on pgBackRest. That's why I said "might probably." It supposedly depends on what's being deleted, and why would other program write something to the dir.

Also, from what I read, Alpine Linux has smaller buffers. That is, under other distros the issue can be reproduced, you just need more files. By the way, `rm` from `coreutils` succeeds under Alpine Linux. Initially it comes with `rm` provided by `busybox`.

Just in case you didn't notice, there's [other issue][a] referenced in my `aports` issue. It has more details.

[a]: https://gitlab.alpinelinux.org/alpine/aports/-/issues/10960

**Comment by dwsteele** [comment]
Agreed, it's annoying, but looks like something we'll have to address, if the kernel can't/wont.

**Comment by x-yuri** [comment]
Okay, I chose CIFS after all. The database takes 112 Mb. The backup takes 44 sec. I guess in other cases it would be faster, but seems tolerable to me.

I investigated the "Directory not empty" issue more and everything points to the fact that it's something end-user software has to deal with. A detailed description [here][a]. Should I file a bug report?

> They seem to have decided that it's a kernel issue.

I'm not sure what they meant by kernel. And I have a vague understanding of what the kernel is. But I'd look for answers in the Samba project. I assume it has some relevant kernel code. Actually, I've [asked][b] today on their mailing list. Maybe we'll get a more authoritative answer.

[a]: https://gist.github.com/x-yuri/aa6db295cd56f757537f756bc390b1df
[b]: https://lists.samba.org/archive/samba/2022-June/241011.html

**Comment by dwsteele** [comment]
> I investigated the "Directory not empty" issue more and everything points to the fact that it's something end-user software has to deal with.

Here's the problem with that: if we are reading the PGDATA directory to build a manifest, what do we do if readdir() produces incorrect results, possibly because Postgres is dropping tables at the same time? For a remove command retries are fine, but for this case there is a major problem. I guess we could keep reading until we get the same result more than once, but wow.

I'm just having trouble with the idea that every app developer has to work around a problem that should be solved in libc or the kernel. At the very least the man page for readir() should clearly state that the results are not reliable if that's going to be the case.

Having griped all that, we are definitely looking into how to deal with this. :)

**Comment by x-yuri** [comment]
> if we are reading the PGDATA directory to build a manifest, what do we do if readdir() produces incorrect results, possibly because Postgres is dropping tables at the same time?

I understand this doesn't involve CIFS. In that case I wasn't able to reproduce the issue with a local filesystem. I guess the reason for this is that CIFS doesn't have a notion of a cursor (which probably involves versioning). That with CIFS it's similar to how pagination is done on the web.

But with a local filesystem I didn't try to delete files at the same time. I'm not sure what the result would be. Considering that there are versioning filesystems, I'm not sure if `readdir()` can always produce a consistent result.

> I'm just having trouble with the idea that every app developer has to work around a problem that should be solved in libc or the kernel.

I don't think it can be solved in libc. In the kernel? Probably in the kernel part of the Samba code if CIFS will add a notion of snapshots (versioning).

Do note that I'm familiar neither with the protocol, nor with the implementation. Those are my speculations.

It all sounds like they sacrificed correctness over simplicity. Which brings to mind the [Worse in better][a] essay. And actually I'm not sure that correctness should be pursued at all cost. It can make things more complex. It also seems like things are moving in that direction too (AI, eventual consistency, design for failure, you name it). At least when you interact over a network, it becomes harder to provide guarantees.

[a]: https://en.wikipedia.org/wiki/Worse_is_better

**Comment by x-yuri** [comment]
Nothing unusual I believe. I tried it on my HP ENVY x360 Convertible laptop (AMD CPU), and on a Hetzner server:

https://www.hetzner.com/dedicated-rootserver (AX41-NVMe)

In case of the Hetzner server that also involved a [Storage Box][a]. I'm not sure if any specifics on it are available though.

Tell me if you need some particular details. Although this looks like something that can be reproduced on any Linux machine. Did you try to reproduce it? It's not just me?

[a]: https://docs.hetzner.com/robot/storage-box/

**Comment by dwsteele** [comment]
We have seen it on a specific commercial appliance. Right now we are trying to determine if this issue is limited to network storage or if it would be possible for it to happen on a local filesystem. If it can happen on a local file system then building a backup manifest is not safe (same would be true for other backup tools and pg_basebackup) because Postgres might be deleting files while we build the manifest.

We have a fix for the remove issue (#1786) but it's just a bandaid and what we don't want to do is mask a larger issue.

**Comment by x-yuri** [comment]
> If it can happen on a local file system then building a backup manifest is not safe (same would be true for other backup tools and pg_basebackup) because Postgres might be deleting files while we build the manifest.

That doesn't tell what you'd expect exactly. Let's say you're reading a directory, you read 3 entries, then some process deletes one of those 3. As a result you get a listing that contains a file that was deleted. I think the most you could expect is a consistent listing at some point in time (a snapshot). And considering that e.g. `ext4` is journaled, I'm not ruling out that that's the case with `ext4`. Again, that's a guess.

Actually, what if ext4 will try to create a snapshot, but if a directory is big enough... At some point abstraction will leak. The question is, "Can it affect pgBackRest in practice?"

Just in case, here's a [test][a] that shows that if you're reading a directory and deleting files, the listing won't be correct (a `readdir()` invocation doesn't verify if the file it returns is still on disk). Particularly, on a local filesystem.

[a]: https://gist.github.com/x-yuri/ec02d1cfe31463ae929974cbdb80c05f

**Comment by dwsteele** [comment]
> the listing won't be correct (a readdir() invocation doesn't verify if the file it returns is still on disk)

This is fine from our perspective. If we later find that the file does not exist then we ignore it. At least while building the backup manifest, which is my primary concern.

**Comment by x-yuri** [comment]
Okay, then [here's][a] a test that deletes a directory with 456975 files under Alpine Linux (with `musl`) on a local filesystem. With `glibc` + CIFS it would fail. With `musl` + CIFS it would fail with 676 files.

[a]: https://gist.github.com/x-yuri/f7c4a373dfc5bb5043aa311cffbce995

**Comment by precision-software** [comment]
From the Posix standard

If a file is removed from or added to the directory after the most recent call to opendir() or rewinddir(), whether a subsequent call to readdir() returns an entry for that file is unspecified.

 

My “educated hunch” says this works on a local Linux filesystem, but it could fail on remote filesystems like NFS and  CIFS.

 

My hunch is based on:

*	Readdir keeps track of its current offset in the directory. It assumes files do not get reassigned to different slots between calls to readdir().
*	Local file systems have “empty” slots so a file can be deleted without immediately rearranging. They also have directory locks so a directory can be rearranged when nobody is reading the directory.
*	If so, readdir() should be able to meet the Posix requirements on a local system.
*	NFS and CIFS protocosl don’t include directory locks or empty slots.<verify!> Deleting a file could rearrange the directory, say by compacting out empty slots.  In this case, readdir() would appear to skip over valid files.

 

Other projects have had problems skipping valid files in readdir(). Their solution was to read the entire directory at once, then process (or delete) the files one by one.

 

unlink() after readdir() leads to incorrect iteration of directory contents · Issue #2528 · emscripten-core/emscripten (github.com) <https://github.com/emscripten-core/emscripten/issues/2528> 

 

Fixed deletion iteration bug in Local::deleteDir() by modevelops · Pull Request #1129 · thephpleague/flysystem (github.com) <https://github.com/thephpleague/flysystem/pull/1129>
